LFS-525: Propogate subject access to nested subjects

acrowthe opened the pull request:
Iteratively search for parent subjects when applying subject restriction pattern
- If a subject is referenced by a second subject, forms for this second subject can now be accessed by users with lfs:subject access to the first subject.

For example, if a user has lfs:subject access to patient p1, and patient p1 has a tumor t1, the user can access forms under both p1 and t1

acrowthe committed:
LFS-525: Propogate subject access to nested subjects
Iteratively search for parent subjects when applying subject restriction pattern

acrowthe commented:
To test: 

1. Run using composum to be able to set up permissions:
`java -jar distribution/target/lfs-*.jar -Dsling.run.modes=dev`

2. Using the admin user, create:
- A patient, tumor and tumor form.
- A new user

3. Using Composum, copy the uuid of the patient
![image](https://user-images.githubusercontent.com/20056751/91323346-c8e45180-e78e-11ea-8350-c1bf95fd1814.png)

4. Using Composum, modify the permissions on the root Form node:
- Delete the allow everyone rule
- Create a rule for the new user, with all permissions and restriction `lfs:subject` with the value being the patients uuid
![image](https://user-images.githubusercontent.com/20056751/91323525-f7622c80-e78e-11ea-84c0-92f92e66dcd1.png)

5. Using the admin user, copy the url to the tumor form.
6. Sign into the new user (eg. using an incognito window to have a second log in) and paste the url of the tumor form.

The form should be loaded.
To verify, follow steps 2, 5 and 6 again with a new patient to verify that a different patient's tumor forms cant be accessed
![image](https://user-images.githubusercontent.com/20056751/91324756-668c5080-e790-11ea-8503-920b4bfb8d0b.png)

IntegralProgrammer commented on modules/permissions/src/main/java/ca/sickkids/ccm/lfs/permissions/internal/SubjectRestrictionPattern.java:
```diff
@@ -57,10 +68,16 @@ public boolean matches(final Tree tree, final PropertyState property)
-            StringUtils.equals(formTree.getProperty("subject").getValue(Type.REFERENCE),
-                this.targetSubject);
+        String subject = formTree.getProperty("subject").getValue(Type.REFERENCE);
+        boolean result = StringUtils.equals(subject, this.targetSubject);
+
+        if (!result) {
+            // Check if the form's subject references the specified subject
+            result = matchesReference(subject);
+        }
```
I think it would be a more elegant solution if you replaced all the selected code with `boolean result = matchesReference(subject)` and then modified the `matchesReference` function so that it first checks if `uuid` is equal to `this.targetSubject` and then moves up the tree searching through parents and parents-of-parents until it finds a node with a `uuid` that is equal to `this.targetSubject`.

acrowthe replied:
Moved string check into matchesReference

acrowthe committed:
LFS-525: Propogate subject access to nested subjects
Address code review

sashaandjic commented:
I have followed the steps and got the correct results. However, while logged as a new user I clicked on the Forms and got this
![image](https://user-images.githubusercontent.com/23243244/91507764-a4c66480-e8a3-11ea-9021-b81853c7ebbe.png)

I was expecting to see the link to the form that the new user has access to

sashaandjic commented:
Also, a new user is unable to create a new form.
![image](https://user-images.githubusercontent.com/23243244/91508048-6bdabf80-e8a4-11ea-95d1-9b019eb5ffc7.png)

acrowthe committed:
LFS-525: Propogate subject access to nested subjects
Fix error creating new form
- SubjectRestrictionPattern was throwing errors when parsing
  incomplete forms which are being created

acrowthe commented:
@sashaandjic
Updated the testing instructions to fix the missing form link issue by adding a permission in step 4
`- Create a rule for everyone with permissions rep:readNodes`

The 500: Server error when creating a new form can also be resolved by adjusting permissions. There were two causes: A bug which has now been fixed and additional permissions that need to be set. LFS-538 was created to add a proper way to set up the permissions, but a workaround was added to step 4.

sashaandjic commented:
The reported problems are fixed. I am not sure about this one: as a new user I can see all subjects, the ones that I have permission to see their form and those whose forms are do not have permission to see.

acrowthe commented:
@sashaandjic 
>as a new user I can see all subjects, the ones that I have permission to see their form and those whose forms are do not have permission to see.

Hiding subjects that a user does not have access to will come in a separate PR: LFS-531 most likely. This change only touches form permissions.

I suspect that improving the error message would be out of scope for this change, since it is a back-end permission change, rather than front end.

sashaandjic commented:
I am having problems creating a new form while logged as a new user. If I try to use an existing subjects it fails with error code 500.
![image](https://user-images.githubusercontent.com/23243244/91758192-95923000-eb9d-11ea-8a53-e8ce4736480e.png)

When I try to create a form with new subjects I get stuck into an infinte Create new subject thing
![image](https://user-images.githubusercontent.com/23243244/91758342-d8540800-eb9d-11ea-9eaa-304d2b401d8e.png)

sashaandjic commented:
The above complaints are not relevant for this PR. I was able to give another user permission to the subject P1 type of patient. The user was able to see the tumor form that admin created involving P1. Also with additional rule user was able to see the form link listed.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/permissions/src/main/java/ca/sickkids/ccm/lfs/permissions/internal/SubjectRestrictionPattern.java:
```diff
@@ -77,4 +88,37 @@ public boolean matches()
+                }
+                subject = this.session.getNodeByIdentifier(nextUuid);
+            }
+        } catch (RepositoryException e) {
+            LOGGER.error("Failed to find subject UUID {}", nextUuid, e);
```
This will usually happen when the user does not have access to the subject node, which is an "expected" scenario.

```suggestion
        } catch (ItemNotFoundException e) {
            LOGGER.debug("Subject UUID {} is inaccessible", nextUuid, e);
        } catch (RepositoryException e) {
            LOGGER.error("Failed to find subject UUID {}", nextUuid, e);
```

acrowthe committed:
LFS-525: Propogate subject access to nested subjects
Adjusted logging when uiser does not have access to a subject
- Do not log an error a user is denied access to a subject
